### PR TITLE
Abstract swarm relationship into SwarmRelation

### PR DESCRIPTION
**Pull request by Yimaha**
This pr abstract majority of "where" clauses and Generic to a single trait, thus allow easier modification in future and readability.

With the change of converting State in Nodes to MonadState, the amount of generic and the relationship between generic is pilling up, This pr would greatly reduce the difficulty for such change and also any future changes.

### CODE REVIEW

File: monad-mock-swarm/src/mock_swarm.rs [hunk 1]
```diff
@@ -5,61 +5,117 @@ use std::{
...
 
-pub struct Node<S, RS, P, LGR, ME, ST, SCT>
-where
-    S: State,
-    RS: RouterScheduler,
-    ME: MockableExecutor<SignatureCollection = SCT>,
-    ST: MessageSignature,
-    SCT: SignatureCollection,
-{
+pub trait SwarmRelation {
```
Comment (Yimaha) [inline monad-mock-swarm/src/mock_swarm.rs]: define relationship between different types within a single trait (by making MonadEvent generic it actually causes a lot of problem, + I gonna replace STATE by forcing MonadState across Swarm so I didn't bother to make it generic)

File: monad-mock-swarm/src/mock_swarm.rs [hunk 2]
```diff
@@ -5,61 +5,117 @@ use std::{
...
+    type TVT: TransactionValidator + Clone;
+    type StateMessage: Deserializable<Self::Message>;
+    type OutboundStateMessage: Serializable<Self::Message>;
+    type Message: Clone + PartialEq + Eq;
+    type LGRCFG;
+    type RSCFG;
+    type MPCFG: Copy;
+}
+// default swarm relation impl
+pub struct MonadSwarm;
```
Comment (Yimaha) [inline monad-mock-swarm/src/mock_swarm.rs]: individual tests are able to define their own customized relationship as needed by creating a struct that impl SwarmRelation

File: monad-mock-swarm/src/mock_swarm.rs [hunk 3]
```diff
@@ -5,61 +5,117 @@ use std::{
...
+    pub executor: MockExecutor<S::STATE, S::RS, S::ME, S::ST, S::SCT>,
+    pub state: S::STATE,
+    pub logger: S::LGR,
+    pub pipeline: S::P,
+    pub pending_inbound_messages: BinaryHeap<Reverse<(Duration, LinkMessage<S::Message>)>>,
     pub rng: ChaCha20Rng,
     pub current_seed: usize,
 }
 
-impl<S, RS, P, LGR, ME, ST, SCT> Node<S, RS, P, LGR, ME, ST, SCT>
```
Comment (Yimaha) [inline monad-mock-swarm/src/mock_swarm.rs:44]: All spots where you used to need to define complex relationships are now grouped together into 1 spot

File: monad-mock-swarm/tests/utils/mod.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    mock::{MockMempool, MockMempoolConfig},
+    mock_swarm::SwarmRelation,
+    transformer::GenericTransformerPipeline,
+};
+use monad_quic::{QuicRouterScheduler, QuicRouterSchedulerConfig};
+use monad_state::{MonadMessage, MonadState, VerifiedMonadMessage};
+use monad_validator::{simple_round_robin::SimpleRoundRobin, validator_set::ValidatorSet};
+use monad_wal::mock::{MockWALogger, MockWALoggerConfig};
+
+pub struct QuicSwarm;
```
Comment (Yimaha) [inline monad-mock-swarm/tests/utils/mod.rs]: File for shared utils across mock-swarm unit test? subject to debait if there is a better design

File: monad-mock-swarm/Cargo.toml [hunk 5]
```diff
@@ -26,20 +26,22 @@ rand_chacha = { workspace = true }
 rayon = { workspace = true }
 tracing = { workspace = true }
 itertools = { workspace = true }
+monad-state = { path = "../monad-state" }
+monad-block-sync = { path = "../monad-block-sync" }
+monad-validator = { path = "../monad-validator" }
+monad-consensus-state = { path = "../monad-consensus-state" }
+
```
Comment (michael-yxchen) [inline monad-mock-swarm/Cargo.toml]: nit: reorganize all `monad-*` dependencies together

File: monad-mock-swarm/Cargo.toml [hunk 6]
```diff
@@ -26,20 +26,22 @@ rand_chacha = { workspace = true }
...
 
 [dev-dependencies]
-monad-block-sync = { path = "../monad-block-sync" }
+monad-gossip = { path = "../monad-gossip" }
+monad-quic = { path = "../monad-quic" }
```
Comment (michael-yxchen) [inline monad-mock-swarm/Cargo.toml]: nit: alphabetically order the crate names?

File: monad-mock-swarm/src/mock_swarm.rs [hunk 7]
```diff
@@ -5,61 +5,117 @@ use std::{
...
+    type TVT = MockValidator;
+    type LGRCFG = MockWALoggerConfig;
+    type RSCFG = NoSerRouterConfig;
+    type MPCFG = MockMempoolConfig;
+    type StateMessage = MonadMessage<Self::ST, Self::SCT>;
+    type OutboundStateMessage = VerifiedMonadMessage<Self::ST, Self::SCT>;
+    type Message = MonadMessage<Self::ST, Self::SCT>;
+}
+
+pub struct Node<S: SwarmRelation> {
```
Comment (michael-yxchen) [inline monad-mock-swarm/src/mock_swarm.rs]: super nit: can we use a different generic name for `SwarmRelation`. I got confused when I saw `S::Message` as `S` used to mean `State`.

File: monad-mock-swarm/src/mock_swarm.rs [hunk 8]
```diff
@@ -176,35 +219,21 @@ enum SwarmEventType {
...
-    ME: MockableExecutor<Event = S::Event, SignatureCollection = SCT>,
-
-    MockExecutor<S, RS, ME, ST, SCT>: Unpin,
-    S::Block: Unpin,
-    Node<S, RS, P, LGR, ME, ST, SCT>: Send,
-    RS::Serialized: Send,
+    MockExecutor<S::STATE, S::RS, S::ME, S::ST, S::SCT>: Unpin,
+    Block<S::SCT>: Unpin,
+    Node<S>: Send,
+    S::Message: Send,
```
Comment (michael-yxchen) [inline monad-mock-swarm/src/mock_swarm.rs]: maybe move this bound to `SwarmRelation`?

File: monad-mock-swarm/src/mock_swarm.rs [hunk 9]
```diff
@@ -5,61 +5,117 @@ use std::{
...
+        Event = MonadEvent<Self::ST, Self::SCT>,
+        SignatureCollection = Self::SCT,
+        Message = Self::StateMessage,
+        OutboundMessage = Self::OutboundStateMessage,
+        Block = FullBlock<Self::SCT>,
+        Config = MonadConfig<Self::SCT, Self::TVT>,
+    >;
+    type ST: MessageSignature + Unpin;
+    type SCT: SignatureCollection + Unpin;
+    type RS: RouterScheduler<M = Self::Message, Serialized = Self::Message, Config = Self::RSCFG>;
```
Comment (michael-yxchen) [inline monad-mock-swarm/src/mock_swarm.rs]: `RS::M` is the typed transport message and `RS::Serialized` is what's sent over the wire. `M` gets serialized to to `Serialized` in `outbound`.

So I think we should put different types on `M` and `Serialized` in this trait.

File: monad-mock-swarm/src/mock_swarm.rs (hunk 2 above)
Comment (michael-yxchen) [inline monad-mock-swarm/src/mock_swarm.rs]: nit: maybe give it a more meaningful name? e.g. `NoserSwarm` to contrast with `QuicSwarm`

File: monad-mock-swarm/Cargo.toml [hunk 10]
```diff
@@ -26,20 +26,22 @@ rand_chacha = { workspace = true }
...
 monad-consensus = { path = "../monad-consensus" }
 monad-consensus-state = { path = "../monad-consensus-state", features = [
     "monad_test",
 ] }
 monad-executor = { path = "../monad-executor", features = ["monad_test"] }
-monad-gossip = { path = "../monad-gossip" }
-monad-quic = { path = "../monad-quic" }
-monad-state = { path = "../monad-state", features = ["monad_test"] }
```
Comment (michael-yxchen) [inline monad-mock-swarm/Cargo.toml:40]: We need this in dev-dependencies so we can run tests from `monad-mock-swarm`

File: monad-mock-swarm/tests/replay_test.rs [hunk 11]
```diff
@@ -45,31 +39,37 @@ type S = MonadState<
...
+    type LGR = LoggerType;
+    type ME = ME;
+    type TVT = MockValidator;
+    type LGRCFG = <Self::LGR as PersistenceLogger>::Config;
+    type RSCFG = NoSerRouterConfig;
+    type MPCFG = MockMempoolConfig;
+    type Message = MonadMessage<Self::ST, Self::SCT>;
+    type StateMessage = MonadMessage<Self::ST, Self::SCT>;
+    type OutboundStateMessage = VerifiedMonadMessage<Self::ST, Self::SCT>;
+}
```
Comment (michael-yxchen) [inline monad-mock-swarm/tests/replay_test.rs]: we can probably move the above type aliases to `impl SwarmRelation for ReplaySwarm`

File: monad-mock-swarm/tests/utils/mod.rs (hunk 4 above)
Comment (michael-yxchen) [reply monad-mock-swarm/tests/utils/mod.rs]: I think this can just go into `src/mock_swarm.rs`

File: monad-testutil/src/swarm.rs [hunk 12]
```diff
@@ -104,49 +104,31 @@ pub fn node_ledger_verification<O: BlockType + PartialEq>(
...
-    RSC: Fn(Vec<PeerId>, PeerId) -> RS::Config,
-
-    LGR::Config: Clone,
-    TVT: TransactionValidator,
+    S: SwarmRelation,
+
+    MockExecutor<S::STATE, S::RS, S::ME, S::ST, S::SCT>: Unpin,
+    Block<S::SCT>: PartialEq + Unpin,
+    Node<S>: Send,
+    S::LGRCFG: Clone,
```
Comment (michael-yxchen) [inline monad-testutil/src/swarm.rs]: we can also consider moving this to `SwarmRelation`. `SwarmRelation::MPCFG` is already `Copy` anyways

File: monad-testutil/examples/logs_gen.rs [hunk 13]
```diff
@@ -6,43 +6,49 @@ use monad_consensus_types::{
...
+    transformer::{GenericTransformer, GenericTransformerPipeline, LatencyTransformer},
 };
-use monad_state::MonadState;
+use monad_state::{MonadMessage, MonadState, VerifiedMonadMessage};
 use monad_testutil::swarm::get_configs;
 use monad_validator::{simple_round_robin::SimpleRoundRobin, validator_set::ValidatorSet};
 use monad_wal::wal::{WALogger, WALoggerConfig};
 
 type SignatureType = NopSignature;
 type SignatureCollectionType = MultiSig<SignatureType>;
```
Comment (michael-yxchen) [inline monad-testutil/examples/logs_gen.rs]: remove these?

File: monad-testutil/src/swarm.rs [hunk 14]
```diff
@@ -176,34 +158,16 @@ pub fn run_nodes_until<S, ST, SCT, RS, RSC, LGR, P, TVT, ME>(
...
-    RS::Serialized: Send,
-
-    RSC: Fn(Vec<PeerId>, PeerId) -> RS::Config,
-
-    LGR::Config: Clone,
-    TVT: Clone,
+    S: SwarmRelation,
+
+    MockExecutor<S::STATE, S::RS, S::ME, S::ST, S::SCT>: Unpin,
+    Block<S::SCT>: PartialEq + Unpin,
```
Comment (michael-yxchen) [inline monad-testutil/src/swarm.rs]: this seems irrelevant now as `SwarmRelation` enforces `State::Block = FullBlock`

File: monad-viz/src/main.rs [hunk 15]
```diff
@@ -72,16 +75,26 @@ type PersistenceLoggerType =
...
+    type LGR = PersistenceLoggerType;
+    type ME = MockableMempoolType;
+    type TVT = MockValidator;
+    type LGRCFG = MockWALoggerConfig;
+    type RSCFG = NoSerRouterConfig;
+    type MPCFG = MockMempoolConfig;
+    type StateMessage = MonadMessage<Self::ST, Self::SCT>;
+    type OutboundStateMessage = VerifiedMonadMessage<Self::ST, Self::SCT>;
+    type Message = MonadMessage<Self::ST, Self::SCT>;
+}
```
Comment (michael-yxchen) [inline monad-viz/src/main.rs]: same here. Can probably move the type aliases into the impl

File: monad-mock-swarm/tests/utils/mod.rs (hunk 4 above)
Comment (Yimaha) [reply monad-mock-swarm/tests/utils/mod.rs]: there is actually a circular dependency issue when its in src/mock_swarm, let me think if there is good alternatives

Comment (Yimaha) [reply monad-mock-swarm/tests/utils/mod.rs]: the solution I ended up doing is migrate the 2 tests from 
`monad_mock_swarm/tests` to `monad_quic/tests`, and thus including all tests that relies on `QuicSwarm`

**Comment by Yimaha** [comment]
@michael-yxchen  comment addressed, awaiting on approval before rebasing

File: monad-mock-swarm/tests/utils/mod.rs (hunk 4 above)
Comment (michael-yxchen) [reply monad-mock-swarm/tests/utils/mod.rs]: My intention was to organize the shared `SwarmRelations` together, either in `src/mock_swarm.rs` or the test util file. I can see the problem now with either approach: the `monad-quic` circular dependency & some external crates are reusing the `NoSerSwarm`. That's rly annoying.

Honestly I don't have any good solutions. Some ideas I have in mind:
- follow the old approach and re-export `NoSerSwarm` in the util function. I drafted this [option](https://github.com/monad-crypto/monad-bft/commit/c2c0483b17b52bc5db8a374b94e31538362b716e.patch) while I was exploring so feel free to reuse 
- implement the `SwarmRelations` in `monad-testutil` so they are in the same place. I think this idea makes more sense as these are indeed test utils.

Comment (michael-yxchen) [reply monad-mock-swarm/tests/utils/mod.rs]: The intention of `monad-mock-swarm` crate is to organize all mockswarm tests together, so I don't think moving two tests to `monad-quic` make sense

Comment (Yimaha) [reply monad-mock-swarm/tests/utils/mod.rs]: if the intention was to keep all related in there then its cool, however I really don't think swarm relation should be part of the test-utils.. tmr i gonna take a look at the patch and update on this

**Comment by Yimaha** [comment]
@michael-yxchen Updated with the patch, and rebased to master

File: monad-mock-swarm/tests/utils/mod.rs (hunk 4 above)
Comment (Yimaha) [reply monad-mock-swarm/tests/utils/mod.rs]: taking the patch :)
